### ISSUE TITLE
Stack overflow with tree-sitter 0.20.8 and 0.20.9 (but works with 0.20.7)

### ISSUE THREAD

**Issue by aibaars** [tree-sitter/tree-sitter#2071]
When testing the tree-sitter-ruby parser on https://github.com/mongodb/mongo-ruby-driver/blob/master/.evergreen/config/standard.yml.erb the parser crashed with a stack overflow. The parser works fine if I downgrade the tree-sitter library to 0.20.7. 

I'm not sure how to debug the problem further. I manually reduced the input to a relatively small test case. Hopefully, it will help you find the cause of the problem:

`Cargo.toml`:
```toml
[package]
name = "test-ruby"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
#tree-sitter = "=0.20.7"
tree-sitter = "=0.20.8"
tree-sitter-ruby = { git = "https://github.com/tree-sitter/tree-sitter-ruby.git", rev = "7a1921bcfd90e3a04c1ad011059087aaf0168dd4" }
```

`src/main.rs`:
```rust
fn main() {
    let ranges = [
        tree_sitter::Range {
            start_byte: 3,
            end_byte: 7,
            start_point: tree_sitter::Point { row: 0, column: 3 },
            end_point: tree_sitter::Point { row: 0, column: 7 },
        },
        tree_sitter::Range {
            start_byte: 157,
            end_byte: 168,
            start_point: tree_sitter::Point { row: 2, column: 2 },
            end_point: tree_sitter::Point { row: 2, column: 13 },
        },
        tree_sitter::Range {
            start_byte: 172,
            end_byte: 179,
            start_point: tree_sitter::Point { row: 3, column: 2 },
            end_point: tree_sitter::Point { row: 3, column: 9 },
        },
    ];
    let source: [u8; 182] = [
        60, 37, 61, 32, 116, 32, 10, 62, 10, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46,
        46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 46, 10, 60, 37, 124, 61,
        32, 102, 111, 111, 32, 100, 111, 32, 10, 62, 10, 60, 37, 124, 32, 101, 110, 100, 32, 10,
        62, 10, 10,
    ];
    let language = tree_sitter_ruby::language();
    let mut parser = tree_sitter::Parser::new();
    parser.set_language(language).unwrap();
    parser.set_included_ranges(&ranges).unwrap();
    println!("start parsing");
    let _tree = parser.parse(&source, None).expect("Failed to parse file");
    println!("end parsing");
}
```

**Comment by aibaars** [comment]
@maxbrunsfeld I believe the diff between 0.20.8 and 0.20.7 is https://github.com/tree-sitter/tree-sitter/compare/db91399ea71aa1e66b521b9b0ef67bcf85ff5e9f...01df16ca9faa1635909ebea242deac200624d9ee . 

Could the changes in https://github.com/tree-sitter/tree-sitter/commit/d223a81b5064587af3a5f61f52d519670ba8995f be the cause? The example input has a syntax error and the Ruby grammar does have some empty tokens.

**Comment by aibaars** [comment]
I guess this an expected effect of https://github.com/tree-sitter/tree-sitter/pull/1783#issuecomment-1181011411 
> If there's an infinite loop, I believe it means that tree-sitter-elm's external scanner is returning an empty token during error recovery. You can detect error recovery in the external scanner by the fact that all tokens are considered valid at once.

**Comment by aibaars** [comment]
@maxbrunsfeld The problematic bit of code in the scanner of tree-sitter-ruby seems to be the following, which appears to yield an infinite stream of `LINE_BREAK` tokens when trying to recover an error at the start of an included range:
https://github.com/tree-sitter/tree-sitter-ruby/blob/master/src/scanner.cc#L205-L209

If I delete that block then things work again for me. However, there must have been a good reason why you added that code. Do you see a good way of fixing this, without deleting the code?

**Comment by ahelwer** [comment]
Possibly related: https://github.com/tree-sitter/tree-sitter/discussions/2073

I do return an empty external token, and it does modify parser state. However, it's far enough down the error recovery chain that I can't detect that we're in error recovery mode in the external scanner by checking whether all tokens are marked valid.

**Comment by aibaars** [comment]
This seems to be fixed by https://github.com/tree-sitter/tree-sitter/pull/1952

**Comment by ahlinc** [comment]
It will be released when #2025 been resolved.
